lib/grandpa/grandpa_test.go

```diff
@@ -1264,3 +1265,160 @@ func TestFinalRoundGaugeMetric(t *testing.T) {
 	gauge := ethmetrics.GetOrRegisterGauge(finalityGrandpaRoundMetrics, nil)
 	require.Equal(t, gauge.Value(), int64(180))
 }
+
+func TestGrandpaServiceCreateJustification_ShouldCountEquivocatoryVotes(t *testing.T) {
+	t.Parallel()
+
+	// setup granpda service
+	gs, st := newTestService(t)
+	addBlocksToState(t, st.Block, 10)
+
+	// create a new fake block to fake authorities vote on
+	previousHash := st.Block.BestBlockHash()
+	previusHead, err := st.Block.BestBlockHeader()
+	require.NoError(t, err)
+
+	bfcNumber := int(previusHead.Number.Int64() + 1)
+
+	d, err := types.NewBabePrimaryPreDigest(0, uint64(bfcNumber), [32]byte{}, [64]byte{}).ToPreRuntimeDigest()
+	require.NoError(t, err)
+	require.NotNil(t, d)
+	digest := types.NewDigest()
+	_ = digest.Add(*d)
+
+	bfcBlock := &types.Block{
+		Header: types.Header{
+			ParentHash: previousHash,
+			Number:     big.NewInt(int64(bfcNumber)),
+			StateRoot:  trie.EmptyHash,
+			Digest:     digest,
+		},
+		Body: types.Body{},
+	}
+
+	bfcHash := bfcBlock.Header.Hash()
+	err = st.Block.AddBlockWithArrivalTime(bfcBlock, time.Now())
+	require.Nil(t, err)
+
+	// create fake authorities
+	fakeAuthorities := make([]*ed25519.Keypair, 0)
+	ed25519Keyring, _ := keystore.NewEd25519Keyring()
+
+	fakeAuthorities = append(fakeAuthorities, ed25519Keyring.Alice().(*ed25519.Keypair))
+	fakeAuthorities = append(fakeAuthorities, ed25519Keyring.Bob().(*ed25519.Keypair))
+	fakeAuthorities = append(fakeAuthorities, ed25519Keyring.Charlie().(*ed25519.Keypair))
+	fakeAuthorities = append(fakeAuthorities, ed25519Keyring.Dave().(*ed25519.Keypair))
+	fakeAuthorities = append(fakeAuthorities, ed25519Keyring.Eve().(*ed25519.Keypair))
+
+	equivocatories := make(map[ed25519.PublicKeyBytes][]*types.GrandpaSignedVote)
+	prevotes := &sync.Map{}
+
+	// voting on
+	for i, v := range fakeAuthorities {
+		vote := &SignedVote{
+			AuthorityID: v.Public().(*ed25519.PublicKey).AsBytes(),
+			Signature:   [64]byte{},
+			Vote: types.GrandpaVote{
+				Hash:   bfcHash,
+				Number: uint32(bfcNumber),
+			},
+		}
+
+		prevotes.Store(i, vote)
```

Comment:
technically the equivocatory votes will all be stored in the `equivocations` map, so for example when we get the first vote from voterA, it'll be stored in the `prevotes` map, but once we get the second conflicting vote from voterA, we will move the previous vote to the equiovcatory map (and store the new vote in it as well)

here it doesn't really matter since it's  a test but I just wanted to mention that

Reply:
ok!